File: ios/Airhop/AirhopVoiceModule.swift
// AirhopVoiceModule: live push-to-talk audio for Airhop.
//
// iOS counterpart of AirhopVoiceModule.kt, with the same contract: the mic
// becomes a stream of AAC-LC frames, and a stream of AAC-LC frames becomes
// sound. No protocol, no routing, no buffering policy — those live in
// voice-capture.ts, flood-router.ts, and voice-player.ts respectively.
//
// Format is fixed by the wire protocol (VoiceBurstCodec.aacLC16kMono = 0x01,
// shared with bitchat): AAC-LC, 16 kHz, mono, 16 kbps, 1024 samples per frame,
// raw frames with no ADTS header. AVAudioConverter produces exactly that.
//
// Threading contract: every exported method returns immediately. Capture runs
// on the audio engine's own render thread and playback on an AVAudioPlayerNode,
// so neither the JS thread nor the UI can be stalled by audio work.
//
// Events emitted to TypeScript:
//   AirhopVoice.frame         { dataBase64 }
//   AirhopVoice.captureError  { message }
import AVFoundation
import Foundation
import React

// MARK: - Constants

private enum VoiceConst {
    // Fixed by the wire format; changing any of these changes what codec byte
    // 0x01 means and breaks both bitchat and older Airhop builds.
    static let sampleRate: Double = 16_000
    static let channels: AVAudioChannelCount = 1
    static let bitRate = 16_000
    // One AAC-LC frame is 1024 samples: 64 ms at 16 kHz, which is the gap the
    // receiver fills with silence when a packet goes missing.
    static let samplesPerFrame: AVAudioFrameCount = 1024
    // Generous ceiling for one encoded frame; ~130 bytes is typical at 16 kbps.
    static let maxEncodedFrameBytes = 1024
    // About two seconds of audio. Past that the sender is outrunning the
    // speaker and the oldest audio is already stale.
    static let maxQueuedFrames = 32
}

private enum VoiceEvent {
    static let frame = "AirhopVoice.frame"
    static let captureError = "AirhopVoice.captureError"
    static let playbackLevel = "AirhopVoice.playbackLevel"
}

/// Loudness of a buffer of float PCM, 0 (silence) to 1 (clipping).
///
/// Plain RMS, deliberately unshaped: speech sits low on a linear scale, so a
/// meter drawn straight from this barely moves. The curve that makes it read
/// well belongs in the UI, where it can be tuned without rebuilding the app.
private func rmsLevel(of buffer: AVAudioPCMBuffer) -> Double {
    guard let channel = buffer.floatChannelData?[0], buffer.frameLength > 0 else { return 0 }
    let count = Int(buffer.frameLength)
    var sum: Float = 0
    for i in 0..<count {
        let sample = channel[i]
        sum += sample * sample
    }
    return Double((sum / Float(count)).squareRoot()).clamped01
}

private extension Double {
    /// Guards the meter against a denormal or an out-of-range sample reaching
    /// the UI as a bar taller than its track.
    var clamped01: Double {
        guard isFinite else { return 0 }
        return Swift.min(1, Swift.max(0, self))
    }
}

// MARK: - Capture generation

/// Which capture a render-thread callback belongs to.
///
/// Removing a tap does not cancel a buffer already handed to it, so a callback
/// from the burst that just ended can still run after the next one has started.
/// Its frames would be attributed to the new burst - the previous talker's last
/// words playing under the new burst's ID on every listener. Every callback
/// proves it is still the current capture before emitting anything.
private final class VoiceCaptureGeneration {
    private let lock = NSLock()
    private var value: UInt64 = 0

    func begin() -> UInt64 {
        lock.lock()
        defer { lock.unlock() }
        value &+= 1
        return value
    }

    func invalidate() {
        lock.lock()
        defer { lock.unlock() }
        value &+= 1
    }

    func isCurrent(_ generation: UInt64) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        return value == generation
    }
}

// MARK: - Module

@objc(AirhopVoiceModule)
final class AirhopVoiceModule: RCTEventEmitter {

    // MARK: Capture state

    // Recreated on every capture, never reused. An engine whose input unit was
    // instantiated against an earlier audio session keeps reporting a dead
    // 0 Hz input and silently fails to enable the microphone, and Airhop hands
    // the session back to playback-only between every burst (the mic button's
    // release path calls setAudioForPlayback). So the second hold and every one
    // after it would capture nothing while reporting success - the burst goes
    // live on the far side and no audio ever follows. bitchat hit the same
    // thing on device and fixed it the same way; see the `engine` comment in
    // their PTTCaptureEngine.swift.
    private var captureEngine = AVAudioEngine()
    // Whether `captureEngine`'s input unit has been instantiated by us. Reading
    // `inputNode` on an engine that was never armed instantiates it against
    // whatever session happens to be active, which is the state this class
    // exists to avoid, so teardown has to know not to touch it.
    private var captureEngineArmed = false
    // Loudness of the most recent block of microphone audio, ridden along on the
    // next frame event rather than sent on its own. The frames already cross the
    // bridge fifteen times a second, which is the rate a meter wants, so this
    // costs no extra traffic and cannot drift out of step with the audio it
    // describes.
    private var captureLevel: Double = 0
    private var captureConverter: AVAudioConverter?
    private var isCapturing = false
    private let captureGeneration = VoiceCaptureGeneration()
    // Serialises capture setup/teardown against the render thread's tap.
    private let captureQueue = DispatchQueue(label: "org.onemindlabs.airhop.voice.capture")

    // MARK: Playback state

    private let playbackEngine = AVAudioEngine()
    private let playerNode = AVAudioPlayerNode()
    private var playbackConverter: AVAudioConverter?
    private var isPlaying = false
    private var queuedFrames = 0
    private let playbackQueue = DispatchQueue(label: "org.onemindlabs.airhop.voice.playback")

    // MARK: RN boilerplate

    @objc override static func requiresMainQueueSetup() -> Bool { false }

    override func supportedEvents() -> [String]! {
        [VoiceEvent.frame, VoiceEvent.captureError, VoiceEvent.playbackLevel]
    }

    // MARK: - Formats

    /// PCM the converters work in: 16 kHz mono float, the engine's native shape.
    private var pcmFormat: AVAudioFormat? {
        AVAudioFormat(
            commonFormat: .pcmFormatFloat32,
            sampleRate: VoiceConst.sampleRate,
            channels: VoiceConst.channels,
            interleaved: false
        )
    }

    /// The compressed format on both sides of the link.
    private var aacFormat: AVAudioFormat? {
        var desc = AudioStreamBasicDescription(
            mSampleRate: VoiceConst.sampleRate,
            mFormatID: kAudioFormatMPEG4AAC,
            mFormatFlags: 0,
            mBytesPerPacket: 0,
            mFramesPerPacket: VoiceConst.samplesPerFrame,
            mBytesPerFrame: 0,
            mChannelsPerFrame: VoiceConst.channels,
            mBitsPerChannel: 0,
            mReserved: 0
        )
        return AVAudioFormat(streamDescription: &desc)
    }

    // MARK: - Capture

    @objc func startCapture(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        captureQueue.async { [weak self] in
            guard let self else { return }
            if self.isCapturing {
                // Idempotent: a double press should not fail a live burst.
                resolve(nil)
                return
            }
            do {
                try self.configureSession()
                try self.beginCapture()
                self.isCapturing = true
                resolve(nil)
            } catch {
                self.isCapturing = false
                self.teardownCapture()
                reject("VOICE_CAPTURE", error.localizedDescription, error)
            }
        }
    }

    @objc func stopCapture(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        captureQueue.async { [weak self] in
            self?.teardownCapture()
            resolve(nil)
        }
    }

    /// Play-and-record with speaker default and ducking, matching the walkie
    /// behaviour the design asks for: other audio dips rather than stops.
    private func configureSession() throws {
        let session = AVAudioSession.sharedInstance()
        try session.setCategory(
            .playAndRecord,
            mode: .voiceChat,
            options: [.duckOthers, .defaultToSpeaker, .allowBluetooth]
        )
        try session.setActive(true, options: [])
    }

    private func beginCapture() throws {
        guard let pcmFormat, let aacFormat else {
            throw VoiceError.format("Unsupported audio format")
        }
        // Fresh engine bound to the session configureSession() just activated.
        // See the `captureEngine` comment for what reuse costs.
        captureEngine = AVAudioEngine()
        captureEngineArmed = true
        let generation = captureGeneration.begin()
        let input = captureEngine.inputNode
        let inputFormat = input.outputFormat(forBus: 0)
        // A dead input reports 0 Hz, and on some routes a plausible rate with no
        // channels; both mean the microphone is not going to produce anything,
        // and failing here lets the caller fall back to a voice note instead of
        // holding a mic that captures silence.
        guard inputFormat.sampleRate > 0, inputFormat.channelCount > 0 else {
            throw VoiceError.format("Microphone unavailable")
        }

        // Two stages: whatever the hardware gives us down to 16 kHz mono, then
        // 16 kHz mono to AAC. AVAudioConverter handles the rate conversion, so
        // this works on hardware that only offers 44.1 or 48 kHz.
        guard let toPCM = AVAudioConverter(from: inputFormat, to: pcmFormat),
              let toAAC = AVAudioConverter(from: pcmFormat, to: aacFormat)
        else {
            throw VoiceError.format("Cannot convert microphone audio")
        }
        toAAC.bitRate = VoiceConst.bitRate
        captureConverter = toAAC

        input.installTap(
            onBus: 0,
            bufferSize: 2048,
            format: inputFormat
        ) { [weak self] buffer, _ in
            self?.handleCapturedBuffer(
                buffer,
                generation: generation,
                toPCM: toPCM,
                toAAC: toAAC,
                pcmFormat: pcmFormat
            )
        }

        captureEngine.prepare()
        try captureEngine.start()
    }

    /// Runs on the audio render thread. Must not allocate more than necessary
    /// and must never block: anything slow here is heard as a dropout.
    private func handleCapturedBuffer(
        _ buffer: AVAudioPCMBuffer,
        generation: UInt64,
        toPCM: AVAudioConverter,
        toAAC: AVAudioConverter,
        pcmFormat: AVAudioFormat
    ) {
        // Belongs to a capture that has already been torn down: its audio is
        // not this burst's to send. See VoiceCaptureGeneration.
        guard captureGeneration.isCurrent(generation) else { return }

        // Resample to 16 kHz mono.
        let ratio = pcmFormat.sampleRate / buffer.format.sampleRate
        let capacity = AVAudioFrameCount(Double(buffer.frameLength) * ratio) + 1024
        guard let pcm = AVAudioPCMBuffer(pcmFormat: pcmFormat, frameCapacity: capacity) else { return }
        var supplied = false
        var error: NSError?
        toPCM.convert(to: pcm, error: &error) { _, status in
            if supplied {
                status.pointee = .noDataNow
                return nil
            }
            supplied = true
            status.pointee = .haveData
            return buffer
        }
        if error != nil || pcm.frameLength == 0 { return }
        // Measured on the resampled 16 kHz mono audio, which is what actually
        // goes out, rather than on whatever shape the hardware handed us.
        captureLevel = rmsLevel(of: pcm)

        // Encode to AAC. The converter emits one packet per 1024 samples, so a
        // buffer that is not a whole number of frames simply carries the
        // remainder into the next call.
        guard let compressed = AVAudioCompressedBuffer(
            format: toAAC.outputFormat,
            packetCapacity: 8,
            maximumPacketSize: VoiceConst.maxEncodedFrameBytes
        ) else { return }

        var pcmSupplied = false
        var encodeError: NSError?
        toAAC.convert(to: compressed, error: &encodeError) { _, status in
            if pcmSupplied {
                status.pointee = .noDataNow
                return nil
            }
            pcmSupplied = true
            status.pointee = .haveData
            return pcm
        }
        if encodeError != nil || compressed.packetCount == 0 { return }

        // Re-checked after the encode: the hold can end while a buffer is in
        // flight, and a frame emitted past that point lands in the next burst.
        guard captureGeneration.isCurrent(generation) else { return }
        emitPackets(from: compressed)
    }

    /// Splits a compressed buffer into its individual AAC packets and sends each
    /// one up as a frame. One packet is one 64 ms frame, which is what the burst
    /// packetizer expects.
    private func emitPackets(from buffer: AVAudioCompressedBuffer) {
        guard let descriptions = buffer.packetDescriptions else {
            // No descriptions means a single packet filling the buffer.
            let data = Data(bytes: buffer.data, count: Int(buffer.byteLength))
            if !data.isEmpty { emitFrame(data) }
            return
        }
        let base = buffer.data.assumingMemoryBound(to: UInt8.self)
        for i in 0..<Int(buffer.packetCount) {
            let description = descriptions[i]
            let offset = Int(description.mStartOffset)
            let size = Int(description.mDataByteSize)
            guard size > 0, offset >= 0 else { continue }
            let data = Data(bytes: base.advanced(by: offset), count: size)
            emitFrame(data)
        }
    }

    private func teardownCapture() {
        isCapturing = false
        captureConverter = nil
        // Invalidated before the engine stops, so a buffer already in flight on
        // the render thread is discarded rather than emitted into whatever
        // burst comes next.
        captureGeneration.invalidate()
        guard captureEngineArmed else { return }
        captureEngineArmed = false
        if captureEngine.isRunning { captureEngine.stop() }
        captureEngine.inputNode.removeTap(onBus: 0)
    }

    // MARK: - Playback

    @objc func startPlayback(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        playbackQueue.async { [weak self] in
            guard let self else { return }
            // One voice at a time: an incoming burst replaces whatever was
            // playing rather than mixing with it.
            self.teardownPlayback()
            do {
                try self.configureSession()
                try self.beginPlayback()
                self.isPlaying = true
                resolve(nil)
            } catch {
                self.teardownPlayback()
                reject("VOICE_PLAYBACK", error.localizedDescription, error)
            }
        }
    }

    @objc func enqueueFrames(
        _ framesBase64: [String],
        resolver resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        playbackQueue.async { [weak self] in
            guard let self else { return }
            // Frames for a burst that already ended. Dropping is correct: late
            // audio is worthless, which is why nothing in this path retransmits.
            guard self.isPlaying else {
                resolve(nil)
                return
            }
            for encoded in framesBase64 {
                guard let frame = Data(base64Encoded: encoded), !frame.isEmpty else { continue }
                // Behind the speaker means the queued audio is already stale.
                if self.queuedFrames >= VoiceConst.maxQueuedFrames { continue }
                self.schedule(frame)
            }
            resolve(nil)
        }
    }

    @objc func stopPlayback(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        playbackQueue.async { [weak self] in
            self?.teardownPlayback()
            resolve(nil)
        }
    }

    private func beginPlayback() throws {
        guard let pcmFormat, let aacFormat else {
            throw VoiceError.format("Unsupported audio format")
        }
        guard let converter = AVAudioConverter(from: aacFormat, to: pcmFormat) else {
            throw VoiceError.format("Cannot decode voice audio")
        }
        playbackConverter = converter

        if playerNode.engine == nil {
            playbackEngine.attach(playerNode)
        }
        playbackEngine.connect(playerNode, to: playbackEngine.mainMixerNode, format: pcmFormat)
        playbackEngine.prepare()
        try playbackEngine.start()
        playerNode.play()
        queuedFrames = 0
    }

    /// Decode one AAC frame and hand the PCM to the player node. The node owns
    /// the pacing: frames play back to back in the order they were scheduled.
    private func schedule(_ frame: Data) {
        guard let converter = playbackConverter,
              let pcmFormat,
              let aacFormat
        else { return }

        guard let compressed = AVAudioCompressedBuffer(
            format: aacFormat,
            packetCapacity: 1,
            maximumPacketSize: frame.count
        ) else { return }
        compressed.byteLength = UInt32(frame.count)
        compressed.packetCount = 1
        frame.withUnsafeBytes { raw in
            guard let base = raw.baseAddress else { return }
            compressed.data.copyMemory(from: base, byteCount: frame.count)
        }
        compressed.packetDescriptions?.pointee = AudioStreamPacketDescription(
            mStartOffset: 0,
            mVariableFramesInPacket: 0,
            mDataByteSize: UInt32(frame.count)
        )

        guard let pcm = AVAudioPCMBuffer(
            pcmFormat: pcmFormat,
            frameCapacity: VoiceConst.samplesPerFrame
        ) else { return }

        var supplied = false
        var error: NSError?
        converter.convert(to: pcm, error: &error) { _, status in
            if supplied {
                status.pointee = .noDataNow
                return nil
            }
            supplied = true
            status.pointee = .haveData
            return compressed
        }
        guard error == nil, pcm.frameLength > 0 else { return }
        // Measured as the audio is handed to the speaker, so the meter shows
        // what is actually being heard rather than what has merely arrived.
        emitPlaybackLevel(rmsLevel(of: pcm))

        queuedFrames += 1
        playerNode.scheduleBuffer(pcm) { [weak self] in
            guard let self else { return }
            self.playbackQueue.async { self.queuedFrames = max(0, self.queuedFrames - 1) }
        }
    }

    private func teardownPlayback() {
        isPlaying = false
        queuedFrames = 0
        if playerNode.engine != nil { playerNode.stop() }
        if playbackEngine.isRunning { playbackEngine.stop() }
        playbackConverter = nil
    }

    // MARK: - Lifecycle

    override func invalidate() {
        captureQueue.sync { teardownCapture() }
        playbackQueue.sync { teardownPlayback() }
        super.invalidate()
    }

    // MARK: - Events

    private func emitFrame(_ frame: Data) {
        guard bridge != nil else { return }
        sendEvent(
            withName: VoiceEvent.frame,
            body: ["dataBase64": frame.base64EncodedString(), "level": captureLevel]
        )
    }

    /// Loudness of audio just handed to the speaker. Unlike capture there is no
    /// existing event to ride along on, so this is its own, at the same fifteen
    /// a second the frames arrive at.
    private func emitPlaybackLevel(_ level: Double) {
        guard bridge != nil else { return }
        sendEvent(withName: VoiceEvent.playbackLevel, body: ["level": level])
    }
}

// MARK: - Errors

private enum VoiceError: LocalizedError {
    case format(String)

    var errorDescription: String? {
        switch self {
        case .format(let message): return message
        }
    }
}

File: android/app/src/main/java/org/onemindlabs/airhop/voice/AirhopVoiceModule.kt
// AirhopVoiceModule: live push-to-talk audio for Airhop.
//
// Turns the microphone into a stream of AAC-LC frames and turns a stream of
// AAC-LC frames back into sound. Nothing else: no protocol, no routing, no
// buffering policy. The burst wire format lives in voice-capture.ts and the
// jitter buffer in voice-player.ts, exactly as the BLE and WiFi modules leave
// routing to TypeScript.
//
// Format is fixed by the wire protocol (VoiceBurstCodec.aacLC16kMono = 0x01,
// shared with bitchat): AAC-LC, 16 kHz, mono, 16 kbps, 1024 samples per frame,
// raw frames with no ADTS header. MediaCodec emits exactly that shape, and the
// decoder is handed the matching AudioSpecificConfig directly rather than
// waiting for one on the wire.
//
// Threading contract: every @ReactMethod returns immediately. Capture runs on
// its own thread and playback on another, so a slow encoder, a busy speaker, or
// a mic grabbed by a phone call can never stall the JS thread or the UI.
//
// Events emitted to TypeScript:
//   AirhopVoice.frame         { dataBase64 }
//   AirhopVoice.captureError  { message }
package org.onemindlabs.airhop.voice

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioRecord
import android.media.AudioTrack
import android.media.MediaCodec
import android.media.MediaCodecInfo
import android.media.MediaFormat
import android.media.MediaRecorder
import android.util.Base64
import android.util.Log
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.WritableNativeMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import java.nio.ByteBuffer
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.atomic.AtomicBoolean

private const val TAG = "AirhopVoiceModule"

// Events emitted to TypeScript.
private const val EVT_FRAME = "AirhopVoice.frame"
private const val EVT_CAPTURE_ERROR = "AirhopVoice.captureError"

// Codec parameters. Fixed by the wire format; changing any of these changes
// what codec byte 0x01 means and breaks both bitchat and older Airhop builds.
private const val SAMPLE_RATE = 16_000
private const val CHANNELS = 1
private const val BIT_RATE = 16_000
private const val AAC_MIME = MediaFormat.MIMETYPE_AUDIO_AAC

// One AAC-LC frame is 1024 samples; at 16-bit mono that is 2048 bytes of PCM.
// Feeding the encoder exactly this much per input buffer keeps one PCM read
// mapped to one output frame, so frame timing matches the 64 ms the receiver's
// silence-fill assumes for a missing packet.
private const val SAMPLES_PER_FRAME = 1024
private const val PCM_BYTES_PER_FRAME = SAMPLES_PER_FRAME * 2 * CHANNELS

// AudioSpecificConfig for AAC-LC / 16 kHz / mono, the decoder's csd-0:
//   audioObjectType 2 (AAC-LC)      = 00010
//   samplingFrequencyIndex 8 (16k)  = 1000
//   channelConfiguration 1 (mono)   = 0001
//   GASpecificConfig padding        = 000
// packed: 00010_1000_0001_000 -> 0x14 0x08
private val AAC_CSD0 = byteArrayOf(0x14, 0x08)

// How long a codec call may wait for a buffer. Short: these run in a loop on a
// dedicated thread, and a long timeout only delays noticing that the loop was
// asked to stop.
private const val CODEC_TIMEOUT_US = 10_000L

// Cap on frames waiting to be played. At 64 ms a frame this is about two
// seconds of audio; past that the sender is outrunning the speaker and the
// oldest audio is already stale, so drop rather than grow without bound.
private const val MAX_QUEUED_FRAMES = 32

class AirhopVoiceModule(
    private val reactContext: ReactApplicationContext,
) : ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String = "AirhopVoice"

    // ---- Capture state -------------------------------------------------------

    private val capturing = AtomicBoolean(false)
    private var captureThread: Thread? = null

    // ---- Playback state ------------------------------------------------------

    private val playing = AtomicBoolean(false)
    private var playbackThread: Thread? = null
    private val frameQueue = LinkedBlockingQueue<ByteArray>()

    private var listenerCount = 0

    // ---- Capture -------------------------------------------------------------

    @ReactMethod
    fun startCapture(promise: Promise) {
        if (capturing.getAndSet(true)) {
            // Already running. Idempotent rather than an error: a double press
            // should not fail a burst that is already live.
            promise.resolve(null)
            return
        }
        val thread = Thread({ runCapture() }, "AirhopVoiceCapture")
        // Audio threads must not be starved by ordinary background work.
        thread.priority = Thread.MAX_PRIORITY
        captureThread = thread
        thread.start()
        promise.resolve(null)
    }

    @ReactMethod
    fun stopCapture(promise: Promise) {
        capturing.set(false)
        // Deliberately not joined: the loop notices the flag within one read
        // (64 ms) and tears itself down. Blocking here would block JS for that
        // long on every release of the mic button.
        captureThread = null
        promise.resolve(null)
    }

    private fun runCapture() {
        var record: AudioRecord? = null
        var encoder: MediaCodec? = null
        try {
            val minBuffer = AudioRecord.getMinBufferSize(
                SAMPLE_RATE,
                AudioFormat.CHANNEL_IN_MONO,
                AudioFormat.ENCODING_PCM_16BIT,
            )
            if (minBuffer <= 0) {
                failCapture("Microphone does not support 16 kHz mono capture")
                return
            }
            record = AudioRecord(
                // VOICE_COMMUNICATION gets the platform's echo cancellation and
                // noise suppression, which is what makes a walkie-talkie usable
                // on speakerphone.
                MediaRecorder.AudioSource.VOICE_COMMUNICATION,
                SAMPLE_RATE,
                AudioFormat.CHANNEL_IN_MONO,
                AudioFormat.ENCODING_PCM_16BIT,
                maxOf(minBuffer, PCM_BYTES_PER_FRAME * 4),
            )
            if (record.state != AudioRecord.STATE_INITIALIZED) {
                failCapture("Microphone unavailable")
                return
            }

            encoder = MediaCodec.createEncoderByType(AAC_MIME).apply {
                configure(
                    MediaFormat.createAudioFormat(AAC_MIME, SAMPLE_RATE, CHANNELS).apply {
                        setInteger(
                            MediaFormat.KEY_AAC_PROFILE,
                            MediaCodecInfo.CodecProfileLevel.AACObjectLC,
                        )
                        setInteger(MediaFormat.KEY_BIT_RATE, BIT_RATE)
                        setInteger(MediaFormat.KEY_MAX_INPUT_SIZE, PCM_BYTES_PER_FRAME * 2)
                    },
                    null,
                    null,
                    MediaCodec.CONFIGURE_FLAG_ENCODE,
                )
                start()
            }

            record.startRecording()
            val pcm = ByteArray(PCM_BYTES_PER_FRAME)
            val info = MediaCodec.BufferInfo()

            while (capturing.get()) {
                val read = record.read(pcm, 0, pcm.size)
                if (read <= 0) {
                    // A negative result is a real error (mic revoked, device
                    // lost); zero just means nothing was ready yet.
                    if (read < 0) {
                        failCapture("Microphone stopped")
                        return
                    }
                    continue
                }
                feedEncoder(encoder, pcm, read)
                drainEncoder(encoder, info)
            }

            // Flush what the encoder is still holding so the tail of the burst
            // is not lost, then let the END packet follow it.
            feedEndOfStream(encoder)
            drainEncoder(encoder, info)
        } catch (e: Exception) {
            Log.w(TAG, "capture failed", e)
            failCapture(e.message ?: "Recording failed")
        } finally {
            capturing.set(false)
            runCatching { record?.stop() }
            runCatching { record?.release() }
            runCatching { encoder?.stop() }
            runCatching { encoder?.release() }
        }
    }

    private fun feedEncoder(encoder: MediaCodec, pcm: ByteArray, length: Int) {
        val index = encoder.dequeueInputBuffer(CODEC_TIMEOUT_US)
        if (index < 0) return // encoder busy; the next read carries on
        val buffer: ByteBuffer = encoder.getInputBuffer(index) ?: return
        buffer.clear()
        buffer.put(pcm, 0, length)
        encoder.queueInputBuffer(index, 0, length, 0L, 0)
    }

    private fun feedEndOfStream(encoder: MediaCodec) {
        val index = encoder.dequeueInputBuffer(CODEC_TIMEOUT_US)
        if (index < 0) return
        encoder.queueInputBuffer(index, 0, 0, 0L, MediaCodec.BUFFER_FLAG_END_OF_STREAM)
    }

    private fun drainEncoder(encoder: MediaCodec, info: MediaCodec.BufferInfo) {
        while (true) {
            val index = encoder.dequeueOutputBuffer(info, 0)
            if (index < 0) return // no output ready, or a format change we ignore
            val output = encoder.getOutputBuffer(index)
            // The codec-config buffer is the AudioSpecificConfig. Receivers
            // rebuild it from the codec byte, so it never goes on the wire.
            val isConfig = info.flags and MediaCodec.BUFFER_FLAG_CODEC_CONFIG != 0
            if (output != null && info.size > 0 && !isConfig) {
                val frame = ByteArray(info.size)
                output.position(info.offset)
                output.get(frame, 0, info.size)
                emitFrame(frame)
            }
            encoder.releaseOutputBuffer(index, false)
            if (info.flags and MediaCodec.BUFFER_FLAG_END_OF_STREAM != 0) return
        }
    }

    // ---- Playback ------------------------------------------------------------

    @ReactMethod
    fun startPlayback(promise: Promise) {
        // One voice at a time: an incoming burst replaces whatever was playing
        // rather than mixing with it.
        stopPlaybackInternal()
        frameQueue.clear()
        playing.set(true)
        val thread = Thread({ runPlayback() }, "AirhopVoicePlayback")
        thread.priority = Thread.MAX_PRIORITY
        playbackThread = thread
        thread.start()
        promise.resolve(null)
    }

    @ReactMethod
    fun enqueueFrames(framesBase64: ReadableArray, promise: Promise) {
        if (!playing.get()) {
            // Frames for a burst that already ended. Dropping is correct: late
            // audio is worthless, which is why there is no retransmit anywhere
            // in this path.
            promise.resolve(null)
            return
        }
        for (i in 0 until framesBase64.size()) {
            val encoded = framesBase64.getString(i) ?: continue
            val frame = runCatching { Base64.decode(encoded, Base64.NO_WRAP) }.getOrNull()
            if (frame == null || frame.isEmpty()) continue
            // Drop the oldest rather than block the bridge when the speaker is
            // behind: the queue holding means the audio in it is already stale.
            while (frameQueue.size >= MAX_QUEUED_FRAMES) frameQueue.poll()
            frameQueue.offer(frame)
        }
        promise.resolve(null)
    }

    @ReactMethod
    fun stopPlayback(promise: Promise) {
        stopPlaybackInternal()
        promise.resolve(null)
    }

    private fun stopPlaybackInternal() {
        playing.set(false)
        // Wake the queue's blocking take so the thread can see the flag and
        // finish. Not joined, for the same reason capture is not.
        frameQueue.offer(ByteArray(0))
        playbackThread = null
    }

    private fun runPlayback() {
        var decoder: MediaCodec? = null
        var track: AudioTrack? = null
        try {
            decoder = MediaCodec.createDecoderByType(AAC_MIME).apply {
                configure(
                    MediaFormat.createAudioFormat(AAC_MIME, SAMPLE_RATE, CHANNELS).apply {
                        setInteger(
                            MediaFormat.KEY_AAC_PROFILE,
                            MediaCodecInfo.CodecProfileLevel.AACObjectLC,
                        )
                        setInteger(MediaFormat.KEY_IS_ADTS, 0)
                        // Handed the config directly: the wire carries raw
                        // frames only, and the codec byte already says what
                        // they are.
                        setByteBuffer("csd-0", ByteBuffer.wrap(AAC_CSD0))
                    },
                    null,
                    null,
                    0,
                )
                start()
            }

            val minBuffer = AudioTrack.getMinBufferSize(
                SAMPLE_RATE,
                AudioFormat.CHANNEL_OUT_MONO,
                AudioFormat.ENCODING_PCM_16BIT,
            )
            track = AudioTrack.Builder()
                .setAudioAttributes(
                    AudioAttributes.Builder()
                        // VOICE_COMMUNICATION routes to the earpiece/speaker the
                        // way a call does and ducks other apps, which is the
                        // behaviour a walkie-talkie wants.
                        .setUsage(AudioAttributes.USAGE_VOICE_COMMUNICATION)
                        .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                        .build(),
                )
                .setAudioFormat(
                    AudioFormat.Builder()
                        .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                        .setSampleRate(SAMPLE_RATE)
                        .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
                        .build(),
                )
                .setBufferSizeInBytes(maxOf(minBuffer, PCM_BYTES_PER_FRAME * 4))
                .setTransferMode(AudioTrack.MODE_STREAM)
                .build()
            track.play()

            val info = MediaCodec.BufferInfo()
            while (playing.get()) {
                // Waits rather than spins: a burst with a gap in it should cost
                // nothing while the gap lasts.
                val frame = frameQueue.take()
                if (frame.isEmpty()) continue // wake-up sentinel from stop
                feedDecoder(decoder, frame)
                drainDecoder(decoder, info, track)
            }
        } catch (e: InterruptedException) {
            Thread.currentThread().interrupt()
        } catch (e: Exception) {
            Log.w(TAG, "playback failed", e)
        } finally {
            playing.set(false)
            runCatching { track?.stop() }
            runCatching { track?.release() }
            runCatching { decoder?.stop() }
            runCatching { decoder?.release() }
            frameQueue.clear()
        }
    }

    private fun feedDecoder(decoder: MediaCodec, frame: ByteArray) {
        val index = decoder.dequeueInputBuffer(CODEC_TIMEOUT_US)
        if (index < 0) return
        val buffer = decoder.getInputBuffer(index) ?: return
        buffer.clear()
        buffer.put(frame)
        decoder.queueInputBuffer(index, 0, frame.size, 0L, 0)
    }

    private fun drainDecoder(
        decoder: MediaCodec,
        info: MediaCodec.BufferInfo,
        track: AudioTrack,
    ) {
        while (true) {
            val index = decoder.dequeueOutputBuffer(info, 0)
            if (index < 0) return
            val output = decoder.getOutputBuffer(index)
            if (output != null && info.size > 0) {
                val pcm = ByteArray(info.size)
                output.position(info.offset)
                output.get(pcm, 0, info.size)
                // Blocking write, but on the playback thread only: this is the
                // speaker setting the pace, which is exactly right.
                track.write(pcm, 0, pcm.size)
            }
            decoder.releaseOutputBuffer(index, false)
        }
    }

    // ---- Lifecycle -----------------------------------------------------------

    override fun invalidate() {
        capturing.set(false)
        stopPlaybackInternal()
        super.invalidate()
    }

    @ReactMethod
    fun addListener(@Suppress("UNUSED_PARAMETER") eventName: String) {
        listenerCount++
    }

    @ReactMethod
    fun removeListeners(count: Int) {
        listenerCount = maxOf(0, listenerCount - count)
    }

    // ---- Events --------------------------------------------------------------

    private fun emitFrame(frame: ByteArray) {
        emitEvent(
            EVT_FRAME,
            WritableNativeMap().apply {
                putString("dataBase64", Base64.encodeToString(frame, Base64.NO_WRAP))
            },
        )
    }

    private fun failCapture(message: String) {
        capturing.set(false)
        emitEvent(EVT_CAPTURE_ERROR, WritableNativeMap().apply { putString("message", message) })
    }

    private fun emitEvent(name: String, params: WritableNativeMap) {
        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            ?.emit(name, params)
    }
}
